helper/src/main/java/com/bloxbean/cardano/yaci/helper/TxSubmissionClient.java

```diff
@@ -34,30 +40,75 @@ public TxSubmissionClient(String host, int port, VersionTable versionTable) {
     private void init() {
         handshakeAgent = new HandshakeAgent(versionTable);
         txSubmissionAgent = new TxSubmissionAgent();
-        n2nClient = new TCPNodeClient(host, port, handshakeAgent, txSubmissionAgent);
+        keepAliveAgent = new KeepAliveAgent();
+
+        n2nClient = new TCPNodeClient(host, port, handshakeAgent, txSubmissionAgent, keepAliveAgent);
 
         handshakeAgent.addListener(new HandshakeAgentListener() {
             @Override
             public void handshakeOk() {
+                log.info("Handshake okay: {}:{}", host, port);
+                keepAliveAgent.sendKeepAlive(1234);
+                txSubmissionAgent.sendNextMessage();
+            }
+        });
+
+        txSubmissionAgent.addListener(new TxSubmissionListener() {
+            @Override
+            public void handleRequestTxs(RequestTxs requestTxs) {
                 txSubmissionAgent.sendNextMessage();
             }
+
+            @Override
+            public void handleRequestTxIdsNonBlocking(RequestTxIds requestTxIds) {
+                submit(false);
+            }
+
+            @Override
+            public void handleRequestTxIdsBlocking(RequestTxIds requestTxIds) {
+                submit(true);
+            }
+
+            private void submit(boolean isBlocking) {
+                if (isBlocking) {
+                    if (txSubmissionAgent.hasPendingTx()) {
+                        txSubmissionAgent.sendNextMessage();
+                    }
+                } else {
+                    txSubmissionAgent.sendNextMessage();
+                }
+            }
+
         });
+
     }
 
-    @Override
-    public void start(Consumer consumer) {
+    public void start() {
         n2nClient.start();
-        txSubmissionAgent.sendNextMessage();
     }
 
-    @Override
     public void shutdown() {
         n2nClient.shutdown();
     }
 
-    @Override
     public boolean isRunning() {
         return n2nClient.isRunning();
     }
 
+    public void submitTxBytes(byte[] txBytes) {
+        var txHash = TxUtil.calculateTxHash(txBytes);
```

Comment:
@satran004 this returns the wrong hash, is there a correct version of this that returns the same hash as `TransactionUtil` from CCL ?

Reply:
@nemo83 You can use CCL's TransactionUtil here.

```
String getTxHash(byte[] transactionBytes)
```